### PR TITLE
WIP: Features/modal windows

### PR DESCRIPTION
**Pull request by danwalmsley**
This PR makes our Window.Owner property behave like in WPF.

This means you can set the Owner of a window then call Show or ShowDialog. The Windows will then behave as stated below (source: https://msdn.microsoft.com/en-us/library/system.windows.window.owner%28v=vs.110%29.aspx)

```
Once this relationship is established, the following behaviors are exhibited:
If an owner window is minimized, all its owned windows are minimized as well.
If an owned window is minimized, its owner is not minimized.
If an owner window is maximized, both the owner window and its owned windows are restored.
An owner window can never cover an owned window.
Owned windows that were not opened using ShowDialog are not modal. The user can still interact with the owner window.
If you close an owner window, its owned windows are also closed.
If an owned window was opened by its owner window using Show, and the owner window is closed, the owned window's Closing event is not raised.
```

### CODE REVIEW

**Comment by grokys** [comment]
cc: @ffquintella as I believe it was you who was having the problem with dialogs.

**Comment by ffquintella** [comment]
Shure it was me. I will clone this as soon as I have some time to test.

**Comment by danwalmsley** [comment]
@grokys if you get chance to try this, it's almost working except if you open an Owned Modal Window (see window page control catalog) then minimise and restore the main window, we end up with the child window with nothing rendered in there

**Comment by danwalmsley** [comment]
A quick resize sorts it, iv tried loads of things but cant see why that happens

**Comment by grokys** [comment]
Ahhh - need to add a reference to Avalonia.ReactiveUI and call `UseReactiveUI`.

**Comment by grokys** [comment]
The window appears empty because when it's restored windows is for some reason clearing its content and the renderer assumes that the content is still there.

I don't have time to test it right now, but I notice in https://msdn.microsoft.com/en-us/library/windows/desktop/ms633591(v=vs.85).aspx it says:

> You must not call SetWindowLong with the GWL_HWNDPARENT index to change the parent of a child window. Instead, use the SetParent function.

Could this be the reason?

**Comment by danwalmsley** [comment]
Child Windows seem to be Windows that are contained within the parent window. This is only setting ownership of the window. So I think it's not related to that.

**Comment by danwalmsley** [comment]
@grokys I didn't see this for some reason

**Comment by grokys** [comment]
> Child Windows seem to be Windows that are contained within the parent window. This is only setting ownership of the window. So I think it's not related to that.

Hmm, ok. The only clue I can give you is that if you remove this call, it works. So that call is changing the behavior of the window to not persist its contents after restore. Is this a documented side-effect of changing the window owner?!

**Comment by danwalmsley** [comment]
I tested this with the new PR, without change in behavior. If you remove the call it works, but there isn't a relationship between the windows anymore so you loose things like minimizing the MainWindow also minimizing the owned window. Is code in Avalonia able to clear the window, or is Windows somehow doing this without us knowing? BTW why don't we render on wm_paint messages? Out of curiosity

**Comment by Sorien** [comment]
Did you try to return 1 on WM_ERASEBKGND?

File: src/Windows/Avalonia.Win32/WindowImpl.cs [hunk 1]
```diff
@@ -817,5 +817,10 @@ public void ShowTaskbarIcon(bool value)
                 UnmanagedMethods.ShowWindow(_hwnd, windowPlacement.ShowCmd);
             }
         }
+
+        public void SetOwner(IWindowImpl owner)
+        {
+            UnmanagedMethods.SetWindowLong(_hwnd, -8, (uint)owner.Handle.Handle);
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/WindowImpl.cs]: Do not do this: https://msdn.microsoft.com/en-us/library/windows/desktop/ms633591(v=vs.85).aspx

>You must not call SetWindowLong with the GWL_HWNDPARENT index to change the parent of a child window. Instead, use the SetParent function.

Comment (danwalmsley) [reply src/Windows/Avalonia.Win32/WindowImpl.cs]: as discussed on Gitter GTK and WPF do this see comments here:
https://github.com/GNOME/gtk/blob/ae7a6773468181b3524c12426f70749682ff0fce/gdk/win32/gdksurface-win32.c#L1800

and also:
https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Window.cs,4257

Please allow me to confirm I hadn't looked at reference source before implementing this, and have only looked now to ensure our compatibility with windows / wpf.

**Comment by danwalmsley** [comment]
Further information: When the parent window is minimized, it receives WM_SIZE, but the owned window doesn't, this causes the parent to redraw.

**Pull request closed without merging**